WT-4639 Use test/format to verify database file portability

lukech opened the pull request:
This PR is to support wiredtiger database file portability testing across different endian platforms. 
- The changes in `evergreen.yml` is used to set up new build variants/tasks infrastructure, and call `./t` program in a for loop to generate wt database files in 10 WT_TEST directories.
- The new script is used to list and dump tables in above generated wt files.
- The evergreen patch build for the changes can be found [here](https://evergreen.mongodb.com/version/5c9df4e4850e610358ee1856).

lukech commented:
@agorrod @keithbostic Could you help to take a look? Thanks.

agorrod commented on test/evergreen.yml:
```diff
@@ -1088,3 +1234,60 @@ buildvariants:
+  batchtime: 1440 # 1 day
+  stepback: false
+  expansions:
+    compile_flags: --ssl MONGO_DISTMOD=ubuntu1604 --release -j3 CCFLAGS="-march=z196 -mtune=zEC12" --variables-files=etc/scons/mongodbtoolchain_gcc.vars
```
I don't understand why this would be necessary? Similarly for goroot and a few others below - do we need them?

lukech replied:
These settings were copied over from `mongodb-mongo-master` s390x build variant, and should be unnecessary to include here. Will remove them to see if a new patch build breaks.

agorrod commented on test/evergreen/dump_wt_datafiles.sh:
```diff
@@ -0,0 +1,89 @@
+		[ "${verbose}" = true ] && echo ${dump}
+	done
+done
+
```
I'd feel better if this test returned an error when no tables were dumped - in case some structural change causes it to no longer find files.

lukech replied:
Will add a counter to track number of tables being dumped out for all data files, and error out if the counter is 0 at the end of the testing.

lukech commented:
@agorrod I pushed a commit to address your comments. Also removed the `distro` settings for the new tasks to let them run on small (default) instances, which proved to be ok by the latest [patch build](https://evergreen.mongodb.com/version/5ca16e27e3c3317253937f65).

keithbostic commented on test/evergreen/dump_wt_datafiles.sh:
```diff
@@ -0,0 +1,100 @@
+
+# Check the existence of 'wt' binary
+wt_binary=$(find ../../ -type f -executable -name wt | grep '.libs' | head -1)
+if [ ! -x "${wt_binary}" ]; then
```
I wouldn't do a `find`, we know where the `wt` binary is built,  and if someone creates a file named `wt` in the future, this will break.

lukech replied:
Fair enough, will replace `find` command with the path of the `wt` binary.

lukech commented on test/evergreen.yml:
```diff
@@ -976,6 +1000,128 @@ tasks:
+            for i in $(seq 10)
+            do
+              ./t -1 -h "WT_TEST.$i"
+            done
```
Thanks. Will add the suggested settings.

agorrod replied:
Did you try using `CONFIG.stress`, and overriding `runs=1` on the command line as we talked about?

lukech replied:
Yes, I tried overriding the `CONFIG.stress` settings with `runs=1` or `-1`, both worked for me. After checking with Keith, the 2 huffman settings in `CONFIG.stress` are not expected for my tests either, so I simply pass the 3 suggested settings in the command line.

agorrod replied:
I'd prefer to stick with the existing testing mechanism for customizing how format is configured.

@keithbostic Do you remember why is huffman disabled usually, and why wouldn't we disable it here?

keithbostic replied:
Huffman makes format run quite slowly (especially Huffman-encoding keys), and it's not being used anywhere as far as I know, so I took it out of the stress test configuration. However, it changes the underlying page format, so I thought it should be included in our endian testing.

keithbostic replied:
It would be easy enough to create `CONFIG.endian`, of course, if that's preferable.

lukech commented on test/evergreen/dump_wt_datafiles.sh:
```diff
@@ -0,0 +1,100 @@
+	for t in ${tables}
+	do
+		echo ${t}
+		dump=$(LD_LIBRARY_PATH="${lib_dir}" ${wt_binary} -h ${d} dump ${t})
```
Will replace dump with verify.

keithbostic replied:
Unless the test is taking too long, we could do both? (I do see both as valuable, although `verify` is more important.)

lukech commented:
@keithbostic I pushed a commit to address your comments. Replaced `dump` with `verify` in the wt command, and updated the script and task names accordingly. 

In the new patch build the `generate` tasks doubled amount of time to finish (1h 14m) likely because of the test/format config settings. I think it's OK since the makespan of these new portability tests is not critical to us.

keithbostic commented:
> In the new patch build the generate tasks doubled amount of time to finish (1h 14m) likely because of the test/format config settings. I think it's OK since the makespan of these new portability tests is not critical to us.

Yes, setting the test time is almost certainly the issue.

Depending on how often the test runs, we could obviously reduce the number of iterations from 10 if we want the test to run more quickly.

As we've discussed, we don't need to run this test to run that much.

lukech commented:
@keithbostic I added table 'dump' back as suggested. The tests are currently scheduled to run once a day.

lukech commented:
@agorrod I pushed a commit to put the test/format settings into `test/format/CONFIG.endian` file. See if it looks ok to you.
